pkg/storage/etcd3/store.go

```diff
@@ -0,0 +1,310 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"errors"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type store struct {
+	client     *clientv3.Client
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+type elemForDecode struct {
+	data []byte
+	rev  uint64
+}
+
+func newStore(c *clientv3.Client, codec runtime.Codec, prefix string) *store {
+	return &store{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (s *store) Backends(ctx context.Context) []string {
+	resp, err := s.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (s *store) Codec() runtime.Codec {
+	return s.codec
+}
+
+func (s *store) Versioner() storage.Versioner {
+	return s.versioner
+}
+
+func (s *store) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractAndResetRevision(s.versioner, obj)
+	if err != nil {
+		return err
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+
+	// Note that if rev = 0, it means the key doesn't exist. In such case, it's equivalent to create.
+	txnResp, err := s.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return storage.NewResourceVersionConflictsError(key, rev)
+	}
+
+	if out != nil {
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (s *store) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
```

Comment:
I agree that we might want to have a quorum-read knob at the level of the class.
However, we don't want to change the interface of this method, because it's just implementation of the existing interface. So if you think we should change it - please open an issue for it.